chore(server): split album update notifications into multiple jobs

uhthomas opened the pull request:
## Description

We would like to move away from the concept of finding and removing pending jobs. The only place this is used is for album update notifications, and this is done so that users who initially uploaded assets to an album will also
receive a notification if someone else then adds assets to the same album. This can also be achieved with a job for each recipient. Multiple jobs also has the advantage that it will scale better for albums with many users, it's possible
to send notifications concurrently, retries are possible without sending duplicate notifications, and it's clear what recipient a job failed for.

Related to #17782

## How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

N/A

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

uhthomas commented:
The assumption that the latest job should include all the recipients should be correct given the code which emits the event finds all users of the album.

https://github.com/immich-app/immich/blob/e822e3eca99285178407e206939070f530b80850/server/src/services/album.service.ts#L169-L175

uhthomas commented:
Would be nice to use "native" notifications instead #17701

jrasm91 commented on server/src/repositories/job.repository.ts:
```diff
@@ -227,19 +230,12 @@ export class JobRepository {
-      return;
-    }
-    try {
+  // deprecated
```
```suggestion
  /** @deprecated */
```

jsdoc comment with `@deprecated` will make it show up with strikethrough and on tooltip

uhthomas commented on server/src/repositories/event.repository.ts:
```diff
@@ -47,7 +47,7 @@ type EventMap = {
 
   // album events
-  'album.update': [{ id: string; recipientIds: string[] }];
+  'album.update': [{ id: string; recipientID: string }];
```
I did search and there isn't really a convention? There are lots of matches for `albumID` for instance.

jrasm91 replied:
The entire server codebase uses Id

uhthomas commented on server/src/services/album.service.ts:
```diff
@@ -170,8 +170,8 @@ export class AlbumService extends BaseService {
-      if (allUsersExceptUs.length > 0) {
-        await this.eventRepository.emit('album.update', { id, recipientIds: allUsersExceptUs });
+      for (const recipientID of allUsersExceptUs) {
+        await this.eventRepository.emit('album.update', { id, recipientID });
```
Yep, that makes sense. Would you like to do that now or later?

jrasm91 replied:
Later is fine

jrasm91 reviewed (approved):
LGTM

alextran1502 reviewed (approved):
Thank you!

alextran1502 commented:
Can you help resolve the conflict? then feel free to merge

uhthomas committed:
chore(server): split album update notifications into multiple jobs

We would like to move away from the concept of finding and removing pending
jobs. The only place this is used is for album update notifications, and this
is done so that users who initially uploaded assets to an album will also
receive a notification if someone else then adds assets to the same album. This
can also be achieved with a job for each recipient. Multiple jobs also has the
advantage that it will scale better for albums with many users, it's possible
to send notifications concurrently, retries are possible without sending
duplicate notifications, and it's clear what recipient a job failed for.
